packages/core/src/browser/shell/tab-bar-toolbar/tab-bar-toolbar-registry.ts

```diff
@@ -103,9 +103,10 @@ export class TabBarToolbarRegistry implements FrontendApplicationContribution {
         }
         const result: Array<TabBarToolbarItem | ReactTabBarToolbarItem> = [];
         for (const item of this.items.values()) {
-            const visible = TabBarToolbarItem.is(item)
+            const visible = TabBarToolbarItem.is(item) && item.command
```

Comment:
I think this overall condition is flawed as I can see the menu also on views, not only editors. I believe the problem is that we do not evaluate the conditions on menu toolbar items (only if they have content) but we do register our menu in `menus-contribution-handler.ts` with a visibility condition. Since this already a rather complex condition, I'd suggest to pull that out into something that is more readable and easier to overwrite, e.g.,:


```
    protected isVisible(item: TabBarToolbarItem | ReactTabBarToolbarItem, widget: Widget): boolean {
        if (AnyToolbarItem.isConditional(item)) {
            if (item.isVisible && !item.isVisible(widget)) {
                return false;
            }
            if (item.when && !this.contextKeyService.match(item.when, widget.node)) {
                return false;
            }
        }
        if (TabBarToolbarItem.is(item) && item.command && !this.commandRegistry.isVisible(item.command, widget)) {
            return false;
        }
        if (MenuToolbarItem.is(item) && !this.hasMenuContent(item, widget)) {
            return false;
        }
        return true;
    }
```

Reply:
A welcome refactoring suggestion.